### PR TITLE
Correctly return attestation data for late requests

### PR DESCRIPTION
**Pull request by prestonvanloon**
Key changes:
- Improve error message when rejecting attestations for violating this condition
- Fix waitToOneThird to use the current slot time for shortcut rather than the head state
- Walk up the canonical chain until we have found a state / root where the slot is appropriate for the request.

Resolves https://github.com/prysmaticlabs/prysm/issues/5164

### CODE REVIEW

File: beacon-chain/rpc/validator/attester.go [hunk 1]
```diff
@@ -84,6 +91,21 @@ func (vs *Server) GetAttestationData(ctx context.Context, req *ethpb.Attestation
...
+	for headState.Slot() > req.Slot {
+		if ctx.Err() != nil {
+			return nil, status.Errorf(codes.Aborted, ctx.Err().Error())
+		}
+		parent := headState.ParentRoot()
+		headRoot = parent[:]
+		headState, err = vs.BeaconDB.State(ctx, parent)
+		if err != nil {
+			return nil, status.Error(codes.Internal, err.Error())
+		}
```
Comment (terencechain) [inline beacon-chain/rpc/validator/attester.go]: what if headState is `nil`? we do delete states and even more with the new state gen service

Comment (prestonvanloon) [reply beacon-chain/rpc/validator/attester.go]: Good point. Do we have state gen service? What do you recommend?

I can return an error as an easy resolution, but I would prefer the more viable / profitable solution

Comment (terencechain) [reply beacon-chain/rpc/validator/attester.go]: Something like this:
https://github.com/prysmaticlabs/prysm/blob/master/beacon-chain/blockchain/head.go#L124

With also an error check at the end:
```go
if headState == nil {
	return status.Error...
}
```

File: beacon-chain/rpc/validator/attester.go [hunk 2]
```diff
@@ -84,6 +92,32 @@ func (vs *Server) GetAttestationData(ctx context.Context, req *ethpb.Attestation
...
+	// In the case that we receive an attestation request after a newer state/block has been
+	// processed, we walk up the chain until state.Slot <= req.Slot to prevent producing an
+	// attestation that violates processing constraints.
+	for headState.Slot() > req.Slot {
+		if ctx.Err() != nil {
+			return nil, status.Errorf(codes.Aborted, ctx.Err().Error())
+		}
+		parent := headState.ParentRoot()
+		headRoot = parent[:]
+		headState, err = vs.BeaconDB.State(ctx, parent)
```
Comment (terencechain) [inline beacon-chain/rpc/validator/attester.go]: Optional comment. I like the pattern below. So when we deprecate `NewStateMgmt`, this becomes much easier to clean up (ie just delete the else block)

```go
if featureconfig.Get().NewStateMgmt {
	headState, err = vs.StateGen.StateByRoot(ctx, parent)
	if err != nil {
		return nil, status.Errorf(codes.Internal, fmt.Sprintf("Failed to generate state: %v", err))
	}
} else {
      headState, err = vs.BeaconDB.State(ctx, parent)
       if err != nil {
			return nil, status.Error(codes.Internal, err.Error())
	}
}
if headState == nil {
	return nil, status.Error(codes.Internal, "Failed to lookup parent state from head.")
}
```

Comment (prestonvanloon) [reply beacon-chain/rpc/validator/attester.go]: Hmm, but what if the db.State returns nil? I would still need to call state gen?

**Pull request merged**